src/udev/net/ethtool-util.c

```diff
@@ -576,8 +576,14 @@ int ethtool_set_glinksettings(int *fd, const char *ifname, struct link_config *l
 
         if (link->advertise) {
                 uint32_t advertise[ETHTOOL_LINK_MODE_MASK_MAX_KERNEL_NU32] = {};
+                Iterator i;
+                void *id;
 
-                advertise[0] = link->advertise;
+                SET_FOREACH(id, link->advertise, i) {
+                        int val = PTR_TO_INT(id);
+
+                        advertise[val / 32] |= (1 << (val % 32));
+                }
```

Comment:
There's no need to use a set. The problem is in the large indices in the definition of `static const char* const advertise_table[]`. Instead of using those enum values as indices, it needs to use the bit shift as the index.

I also notice that the kernel now has a bunch more of negotiation modes which we don't support:
```C
        ETHTOOL_LINK_MODE_40000baseKR4_Full_BIT = 23,
        ETHTOOL_LINK_MODE_40000baseCR4_Full_BIT = 24,
        ETHTOOL_LINK_MODE_40000baseSR4_Full_BIT = 25,
        ETHTOOL_LINK_MODE_40000baseLR4_Full_BIT = 26,
        ETHTOOL_LINK_MODE_56000baseKR4_Full_BIT = 27,
        ETHTOOL_LINK_MODE_56000baseCR4_Full_BIT = 28,
        ETHTOOL_LINK_MODE_56000baseSR4_Full_BIT = 29,
        ETHTOOL_LINK_MODE_56000baseLR4_Full_BIT = 30,
        ETHTOOL_LINK_MODE_25000baseCR_Full_BIT  = 31,
        ETHTOOL_LINK_MODE_25000baseKR_Full_BIT  = 32,
        ETHTOOL_LINK_MODE_25000baseSR_Full_BIT  = 33,
        ETHTOOL_LINK_MODE_50000baseCR2_Full_BIT = 34,
        ETHTOOL_LINK_MODE_50000baseKR2_Full_BIT = 35,
        ETHTOOL_LINK_MODE_100000baseKR4_Full_BIT        = 36,
        ETHTOOL_LINK_MODE_100000baseSR4_Full_BIT        = 37,
        ETHTOOL_LINK_MODE_100000baseCR4_Full_BIT        = 38,
        ETHTOOL_LINK_MODE_100000baseLR4_ER4_Full_BIT    = 39,
        ETHTOOL_LINK_MODE_50000baseSR2_Full_BIT         = 40,
        ETHTOOL_LINK_MODE_1000baseX_Full_BIT    = 41,
        ETHTOOL_LINK_MODE_10000baseCR_Full_BIT  = 42,
        ETHTOOL_LINK_MODE_10000baseSR_Full_BIT  = 43,
        ETHTOOL_LINK_MODE_10000baseLR_Full_BIT  = 44,
        ETHTOOL_LINK_MODE_10000baseLRM_Full_BIT = 45,
        ETHTOOL_LINK_MODE_10000baseER_Full_BIT  = 46,
        ETHTOOL_LINK_MODE_2500baseT_Full_BIT    = 47,
        ETHTOOL_LINK_MODE_5000baseT_Full_BIT    = 48,
```

So instead of switching to a Set, please:
1. keep the changes of `typedef enum NetDevAdvertise` from this PR
2. change the type of `link_config->advertise` to `uint64_t`
3. use `1 << ...`  when storing the parsed advertise modes in `link_config->advertise`
4. add any modes that the kernel knows to our table

Reply:
Oh, I see that parts of this are implemented in  #10819, I'll comment there.

Reply:
Not really. I don't agree with you. I see it as a individual items which can;t be duplicated. Now it all about get those unique values and apply to kernel. why make it 64 bit. why change later ? this way it does not matter at all .

Reply:
BTW I opened this PR just because I was is CC . I  thought that was my mistake and I should fix this.